File: nextjs/src/type/customer.ts
// customer.ts
export interface CustomerAttributeBase {
  name: string;
  name_ruby: string;
  mail_address: string;
  phone_number: string;
  sex: 'male' | 'female' | 'unknown';
  postal_code: string;
  prefecture: string;
  street: string;
  address: string;
  building: string;
}

export interface CustomerAttribute extends CustomerAttributeBase {
  id: number;
  customer_id: number;
}

export interface CustomerBase {
  customer_attributes: CustomerAttribute[];
}

export interface CustomerCreate extends CustomerBase {}

export interface Customer extends CustomerBase {
  id: number;
  store_id: number;
  created_at: string | Date;
  updated_at: string | Date;
}

File: fastapi/app/schemas/customer.py
# schemas/customer.py
from pydantic import BaseModel
from typing import Optional, List
from enum import Enum
from datetime import datetime

class SexEnum(str, Enum):
    male = "male"
    female = "female"
    unknown = "unknown"

class CustomerAttributeBase(BaseModel):
    name: str
    name_ruby: str
    mail_address: str
    sex: SexEnum
    phone_number: str
    postal_code: str
    prefecture: str
    street: str
    address: str
    building: str

class CustomerAttributeResponseSchema(BaseModel):
    name: str
    name_ruby: str
    mail_address: str
    sex: SexEnum
    phone_number: str
    postal_code: str
    prefecture: str
    street: str
    address: str
    building: str
    created_at: datetime
    updated_at: datetime

    class Config:
        orm_mode = True

class CustomerCreateSchema(BaseModel):
    customer_attributes: CustomerAttributeBase

class CustomerResponseSchema(BaseModel):
    id: int
    store_id: int
    customer_attributes: List[CustomerAttributeResponseSchema]
    created_at: datetime
    updated_at: datetime

    class Config:
        orm_mode = True

class CustomerUpdateSchema(BaseModel):
    customer_attributes: CustomerAttributeBase

    class Config:
        orm_mode = True